src/buf/fixed/allocator.rs

```diff
@@ -0,0 +1,214 @@
+use std::{alloc::Layout, cell::RefCell, io, ptr::NonNull, rc::Rc};
+
+use buddy_alloc::{buddy_alloc::BuddyAlloc, BuddyAllocParam};
+use libc::iovec;
+
+use crate::runtime::CONTEXT;
+
+use super::{buffers::AllocatableBuffers, handle::AllocatedBuf, FixedBuf, FixedBuffers};
+
+/// An allocateable I/O buffer heap pre-registered with the kernel.
+///
+/// `FixedBufAllocator` allows the application to manage a buffer heap
+/// allocated in memory, that can be registered in the current `tokio-uring`
+/// context using the [`register`] method. Unlike [`FixedBufRegistry`],
+/// individual buffers are not retrieved by index; instead, an available
+/// buffer matching a specified capacity can be retrieved with the [`allocate_fixed`]
+/// method. This allows some flexibility in managing buffers without pre-selecting
+/// the desired capacities required, at the expense of some bookeeping overhead.
+///
+/// A `FixedBufAllocator` value is a lightweight handle for a buffer heap.
+/// Cloning of a `FixedBufAllocator` creates a new reference to
+/// the same buffer heap.
+///
+/// The buffers of the collection are not deallocated until:
+/// - all `FixedBufAllocator` references to the heap have been dropped;
+/// - The `tokio-uring` [`Runtime`] the buffers are registered with
+///   has been dropped.
+///
+/// [`register`]: Self::register
+/// [`allocate_fixed`]: Self::allocate_fixed
+/// [`Runtime`]: crate::Runtime
+#[derive(Clone)]
+pub struct FixedBufAllocator {
+    inner: Rc<RefCell<Inner>>,
+}
+
+impl FixedBufAllocator {
+    /// Creates a new buffer heap of the requested size.
+    pub fn new(size: usize) -> io::Result<Self> {
+        // the buddy allocator will use some of the space in the buffer for metadata
+        // and so the layout.size() will be greater than allocator.available_bytes()
+        let layout = Layout::from_size_align(size, 4096)
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
+        let heap = unsafe {
+            let data = std::alloc::alloc(layout) as *mut u8;
```

Comment:
Did you consider anonymous mmap here instead.

Reply:
I'm not sure I know what the advantages/disadvantages are here. What would using the anonymous mmap gain us?